### PR TITLE
new template for creating an HDInsight cluster with an empty edgenode

### PR DESCRIPTION
**Pull request by mumian**
### Contributing guide
https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md

### Changelog
*
*
*
*
*

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@mumian__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (jgao). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by mumian** [comment]
I have fixed the error.  Can you check again?

**Comment by mumian** [comment]
Using Gen-unique and gen-password failed the test.  I had to put a values in the parameters file.  

Now the template has passed the test.  What is the next step?

**Comment by mumian** [comment]
#sign-off

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        "Japan East",
+        "Japan West",
+        "Australia East",
+        "Australia Southeast"
+      ],
+      "metadata": {
+        "description": "The location where all azure resources will be deployed."
+      }
+    },
+    "clusterType": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: remove this and just use a literal value

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        "hadoop",
+        "hbase",
+        "storm",
+        "spark"
+      ],
+      "metadata": {
+        "description": "The type of the HDInsight cluster to create."
+      }
+    },
+    "clusterWorkerNodeCount": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: remove this and just use a literal value

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+      "type": "int",
+      "defaultValue": 2,
+      "metadata": {
+        "description": "The number of nodes in the HDInsight cluster."
+      }
+    },
+    "edgeNodeSize": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: remove this and just use a literal value

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+          "Standard_A2",
+          "Standard_A3",
+          "Standard_A4",
+          "Standard_D1",
+          "Standard_D2",
+          "Standard_D3",
+          "Standard_D4"
+      ]
+    },
+    "edgeNodePrefix": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: remove this and just use a literal value

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        "type": "string",
+        "metadata": {
+            "description": "Edgenode prefix"
+        },
+        "defaultValue": "new"
+    }
+  },
+  "variables": {
+    "defaultApiVersion": "2015-05-01-preview",
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: use literal values for apiVersion so schema validation can be done...

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        "description": "These credentials can be used to remotely access the cluster."
+      }
+    },
+    "sshPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "The password must be at least 10 characters in length and must contain at least one digit, one non-alphanumeric character, and one upper or lower case letter."
+      }
+    },
+    "location": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: location is ambiguous, use clusterLocation or resourceLocation as appropriate

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,227 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "clusterName": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: is this globally unique (should note that in the description)?  better yet can we generate this w/ uniqueString()?

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+              "restAuthCredential.isEnabled": true,
+              "restAuthCredential.username": "[parameters('clusterLoginUserName')]",
+              "restAuthCredential.password": "[parameters('clusterLoginPassword')]"
+            }
+          }
+        },
+        "storageProfile": {
+          "storageaccounts": [
+            {
+              "name": "[concat(variables('clusterStorageAccountName'),'.blob.core.windows.net')]",
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: see #9 in the BP guide

File: 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json [hunk 9]
```diff
@@ -0,0 +1,36 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "clusterName": {
+          "value": "myhdicluster"
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json]: GEN-UNIQUE

File: 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        },
+        "clusterLoginUserName": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json]: GEN-USER

File: 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+            "value": "admin"
+        },
+        "clusterLoginPassword": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json]: GEN-PASSWORD

File: 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+          "value": "Pass@word123"
+        },
+        "sshUserName": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json]: GEN-USER

File: 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            "value": "sshuser"
+        },
+        "sshPassword": {
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.parameters.json]: GEN-PASSWORD

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+                    "name": "edgenode",
+                    "targetInstanceCount": 1,
+                    "hardwareProfile": {
+                        "vmSize": "[parameters('edgeNodeSize')]"
+                    }
+                }]
+            },
+            "installScriptActions": [{
+                "name": "[concat('emptynode','-' ,uniquestring(variables('applicationName')))]",
+                "uri": "https://raw.githubusercontent.com/hdinsight/Iaas-Applications/master/EmptyNode/scripts/EmptyNodeSetup.sh",
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#samples-that-contain-extra-artifacts-custom-scripts-nested-templates-etc

for how to handle artifacts

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json (hunk 7 above)
Comment (mumian) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: I have tried GEN-UNIQUE.  The template won't pass the validation process.

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json (hunk 1 above)
Comment (mumian) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: Do you mean using a variable instead of a parameter to keep the template simple?

Comment (mumian) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: Never mind.  I have this fixed.

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json (hunk 7 above)
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: What error are you getting?  Other templates seem to be making it through CI generating param values...

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+  "variables": {
+    "clusterStorageAccountName": "[concat(parameters('clusterName'),'store')]",
+    "applicationName": "new-edgenode"
+  },
+  "resources": [
+    {
+      "name": "[variables('clusterStorageAccountName')]",
+      "type": "Microsoft.Storage/storageAccounts",
+      "location": "East US",
+      "apiVersion": "2015-05-01-preview",
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: use a non-preview version, make sure it matches the use in the ref function below

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+              "restAuthCredential.isEnabled": true,
+              "restAuthCredential.username": "[parameters('clusterLoginUserName')]",
+              "restAuthCredential.password": "[parameters('clusterLoginPassword')]"
+            }
+          }
+        },
+        "storageProfile": {
+          "storageaccounts": [
+            {
+              "name": "[replace(replace(reference(resourceId('Microsoft.Storage/storageAccounts', variables('clusterStorageAccountName')), '2016-01-01').primaryEndpoints.blob,'https://',''),'/','')]",
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: api version here should match what you're using to define the template resource

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+              "isDefault": true,
+              "container": "[parameters('clusterName')]",
+              "key": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('clusterStorageAccountName')), '2015-05-01-preview').key1]"
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: api version here should match what you're using to define the template resource

File: 101-hdinsight-linux-with-edge-node/azuredeploy.json [hunk 18]
```diff
@@ -88,7 +87,7 @@
               "name": "[replace(replace(reference(resourceId('Microsoft.Storage/storageAccounts', variables('clusterStorageAccountName')), '2016-01-01').primaryEndpoints.blob,'https://',''),'/','')]",
               "isDefault": true,
               "container": "[parameters('clusterName')]",
-              "key": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('clusterStorageAccountName')), '2016-01-01').key1]"
+              "key": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('clusterStorageAccountName')), '2015-05-01-preview').key1]"
```
Comment (bmoore-msft) [inline 101-hdinsight-linux-with-edge-node/azuredeploy.json]: I think you want to use the same apiVersion as the one you're provisioning... there might be a breaking change between those versions and listKeys so something like:

listKeys(resourceId('Microsoft.Storage/storageAccounts', parameters('clusterStorageAccountName')), '2016-01-01').keys[0].value)

**Comment by mumian** [comment]
Brian: thank you for all your help.  The template has passed the check.

**Pull request merged**